src/FlatpakBundleFile.vala

```diff
@@ -0,0 +1,309 @@
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+* General Public License for more details.
+*
+* You should have received a copy of the GNU General Public
+* License along with this program; if not, write to the
+* Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+* Boston, MA 02110-1301 USA
+*
+*/
+
+public class Sideload.FlatpakBundleFile : Object {
+    public File file { get; construct; }
+
+    public string? install_size { get; private set; default = null; }
+    public bool already_installed { get; private set; default = false; }
+    public bool has_remote { get; private set; default = false; }
+
+    private string? appdata_name = null;
+
+    public signal void installation_failed (GLib.Error details);
+    public signal void installation_succeeded ();
+    public signal void details_ready ();
+
+    private Flatpak.BundleRef bundle = null;
+
+    private static Flatpak.Installation? installation;
+
+    static construct {
+        try {
+            installation = new Flatpak.Installation.user ();
+        } catch (Error e) {
+            warning (e.message);
+        }
+    }
+
+    public FlatpakBundleFile (File file) {
+        Object (file: file);
+    }
+
+    construct {
+        try {
+            bundle = new Flatpak.BundleRef (file);
+            install_size = GLib.format_size (bundle.get_installed_size ());
+        } catch (Error e) {
+            warning (e.message);
+        }
+    }
+
+    private async string? get_id () {
+        return ((Flatpak.Ref) bundle).get_name ();
+    }
+
+    public async string? get_name () {
+        // Application name from AppData is preferred
+        if (appdata_name != null) {
+            return appdata_name;
+        }
+
+        // Otherwise, fallback to the app id
+        return yield get_id ();
+    }
+
+    private async string? get_branch () {
+        return ((Flatpak.Ref) bundle).get_branch ();
+    }
+
+    private async void dry_run (Cancellable? cancellable) throws GLib.Error {
+        if (installation == null) {
+            throw new IOError.FAILED (_("Did not find suitable Flatpak installation."));
+        }
+
+        var added_remotes = new Gee.ArrayList<string> ();
+
+        try {
+            var flatpak_id = yield get_id ();
+
+            // get_appstream () only returns the bytes form the appstream file.
+            // so we create a temporary file to parse it.
+            var appstream_file = GLib.File.new_for_path (file.get_path () + ".appstream");
+            try {
+                // get the appstream info from inside the bundle
+                var stream = appstream_file.create (FileCreateFlags.NONE);
+                yield stream.write_bytes_async (bundle.get_appstream ());
+
+                // If we can't find the app by ID in the appstream data, try with a .desktop suffix
+                if (!parse_xml (appstream_file, flatpak_id)) {
```

Comment:
Is this uncompressed appstream data? If so, would it be possible to parse it with something like [`AppStream.Metadata.parse_bytes`](https://valadoc.org/appstream/AppStream.Metadata.parse_bytes.html) ?

Reply:
No, it's a compressed appstream file.